woocommerce/wc-admin
# Extended task list refactor

Related issue [#5242](https://github.com/woocommerce/woocommerce-admin/issues/5242) and [PR #5794](https://github.com/woocommerce/woocommerce-admin/pull/5794)

This PR addresses the following refactors:
* the extended task list hooking.
* task list element handling.
* naming modifications.
* infinite `home` screen looping when there is more than 1 task. 

**Summary**
* It should allow 3pds to create setup tasks for their extensions with the following items: task name (with link to an extension page), time to complete the tasks, and a dismiss feature
* If a merchant completes the extended task list and then installs an extension that has a task associated, then the task list will be displayed again with that extension.

### Screenshots
![screenshot-one wordpress test-2020 12 02-17_01_48](https://user-images.githubusercontent.com/1314156/100925052-3c99a700-34c0-11eb-9a32-f0eda5c7c75a.png)

### Detailed test instructions:
1. Run `npm run example -- --ext=add-task`.
2. Activate the plugin `WooCommerce Admin Add Task Example`.
3. Be sure the setup task list is not hidden
```
UPDATE `wp_options` SET `option_value` = 'no' WHERE `option_name` = 'woocommerce_task_list_hidden';
```
4. Go to the `Home` screen.
5. Verify the task list `Extensions setup` is visible.
6. Try the `Dismiss` button and undo the dismission. Verify everything is working as expected.

7. Hide the setup list (`Finish store setup`) and verify the other list is still there.
8. Hide the `Extensions setup` list and verify that it's not visible anymore. 
![screenshot-one wordpress test-2020 12 02-18_49_37](https://user-images.githubusercontent.com/1314156/100935763-470f6d00-34cf-11eb-98a2-b83fc0dac808.png)
9. Try adding another task * and verify the list now is visible again.
10. Dismiss every extended task.
11. The extended task list won't be visible anymore. You can see every dismissed task by executing an SQL statement like:
```
SELECT * FROM `wp_options` WHERE `option_name` = 'woocommerce_task_list_dismissed_tasks';
``` 

_* I did that in the branch `add/5833_update_test` so you can check out that branch and run `npm run example -- --ext=add-task-2`, and activate the plugin `WooCommerce Admin Add Task Example 2`._



### Update
**Extra test instructions** :
1. Go to the orders page (`/wp-admin/edit.php?post_type=shop_order`).
2. Go to Settings > General > Help (accordion) > Setup wizard
3. Hide the setup task list and verify the button under the title `Task list` is `Enable`.
4. Enable the setup task list and verify that now the setup task list is visible.
5. Do the same with the extended task list.
